{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# train  LG GB SVM-BRF models on imbalanced/balanced ClinVar data with 29 features\n",
    "\n",
    "Method:  \n",
    "Sklearn data preprocessing pipeline,   \n",
    "joblib save model, save preprocessor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from joblib import dump, load\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# extract  features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Users/nancy/Desktop/RS_projects/data/03_ML/01_featureV1/01_train/modelV2/01_ALLfeatures_extracted_addCLNSIG.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 111\u001b[0m\n\u001b[1;32m    108\u001b[0m head_line \u001b[39m=\u001b[39m \u001b[39mfloat\u001b[39m(\u001b[39m'\u001b[39m\u001b[39minf\u001b[39m\u001b[39m'\u001b[39m)                                            \u001b[39m#  as infinity integer\u001b[39;00m\n\u001b[1;32m    110\u001b[0m \u001b[39mwith\u001b[39;00m gzip\u001b[39m.\u001b[39mopen(in_file) \u001b[39mas\u001b[39;00m in_f:                                    \u001b[39m# read compressed file\u001b[39;00m\n\u001b[0;32m--> 111\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39;49m(out_file,\u001b[39m'\u001b[39;49m\u001b[39mw\u001b[39;49m\u001b[39m'\u001b[39;49m) \u001b[39mas\u001b[39;00m out_f:                               \u001b[39m# write new file.csv\u001b[39;00m\n\u001b[1;32m    112\u001b[0m         \n\u001b[1;32m    113\u001b[0m         \u001b[39m### write the head of features outside CSQ and CSQ\u001b[39;00m\n\u001b[1;32m    114\u001b[0m         out_f\u001b[39m.\u001b[39mwrite(\u001b[39m'\u001b[39m\u001b[39m;\u001b[39m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39mjoin(features) \u001b[39m+\u001b[39m\u001b[39m'\u001b[39m\u001b[39m;\u001b[39m\u001b[39m'\u001b[39m)                        \n\u001b[1;32m    115\u001b[0m         out_f\u001b[39m.\u001b[39mwrite(\u001b[39m'\u001b[39m\u001b[39m;\u001b[39m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39mjoin(csq_name_split)\u001b[39m+\u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)                  \n",
      "File \u001b[0;32m~/miniconda3/envs/bioinfo/lib/python3.10/site-packages/IPython/core/interactiveshell.py:282\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    275\u001b[0m \u001b[39mif\u001b[39;00m file \u001b[39min\u001b[39;00m {\u001b[39m0\u001b[39m, \u001b[39m1\u001b[39m, \u001b[39m2\u001b[39m}:\n\u001b[1;32m    276\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    277\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mIPython won\u001b[39m\u001b[39m'\u001b[39m\u001b[39mt let you open fd=\u001b[39m\u001b[39m{\u001b[39;00mfile\u001b[39m}\u001b[39;00m\u001b[39m by default \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    278\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    279\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39myou can use builtins\u001b[39m\u001b[39m'\u001b[39m\u001b[39m open.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    280\u001b[0m     )\n\u001b[0;32m--> 282\u001b[0m \u001b[39mreturn\u001b[39;00m io_open(file, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/nancy/Desktop/RS_projects/data/03_ML/01_featureV1/01_train/modelV2/01_ALLfeatures_extracted_addCLNSIG.csv'"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "\"\"\"1.for outside CSQ: get feature names \"\"\"\n",
    "\n",
    "features = ['CLNSIG','CADD','AF_ESP','AF_EXAC','AF_TGP','Frq','GNOMADAF','GNOMADAF_popmax','Hom','ORIGIN','SPIDEX','SWEGENAF'] ## keep CLNSIG\n",
    "\n",
    "## get features' value start site\n",
    "feature_value_start={}\n",
    "for i in features:\n",
    "    feature_value_start[i]= (len(i) +1)\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"2.for outside CSQ,function: extract features' value \"\"\"\n",
    "def extract_feature(features,feature_value_start,info):\n",
    "    features_outside = []\n",
    "\n",
    "    if any('SWEGENAAC_Hom=' in str for str in info):             # remove 'SWEGENAAC_Hom=' in advance, avoid mess 'HOM' later\n",
    "        match = [s for s in info if 'SWEGENAAC_Hom=' in s]       # match is a list, [0] to get str item\n",
    "        try:\n",
    "            info.remove(match[0])\n",
    "        except ValueError:\n",
    "            pass\n",
    "\n",
    "\n",
    "    for i in features:\n",
    "        prefix = (i+'=')   \n",
    "        if any (prefix in str for str in info):                  # check for the presence 'score' in INFO:\n",
    "            matching = [s for s in info if prefix in s]          # get the item containing 'score':\n",
    "            value_start =  feature_value_start[i]          \n",
    "            features_outside.append(matching[0][value_start:])   # get the value\n",
    "        else:\n",
    "            features_outside.append('')                          # if not exist, add empty\n",
    "\n",
    "    return features_outside                                      # 2D list\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"3.for CSQ: get 81 fields names\"\"\"\n",
    "\n",
    "csq_name='Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|CANONICAL|TSL|APPRIS|CCDS|ENSP|SWISSPROT|TREMBL|UNIPARC|UNIPROT_ISOFORM|REFSEQ_MATCH|SOURCE|REFSEQ_OFFSET|GIVEN_REF|USED_REF|BAM_EDIT|SIFT|PolyPhen|DOMAINS|HGVS_OFFSET|MOTIF_NAME|MOTIF_POS|HIGH_INF_POS|MOTIF_SCORE_CHANGE|TRANSCRIPTION_FACTORS|MES-NCSS_downstream_acceptor|MES-NCSS_downstream_donor|MES-NCSS_upstream_acceptor|MES-NCSS_upstream_donor|MES-SWA_acceptor_alt|MES-SWA_acceptor_diff|MES-SWA_acceptor_ref|MES-SWA_acceptor_ref_comp|MES-SWA_donor_alt|MES-SWA_donor_diff|MES-SWA_donor_ref|MES-SWA_donor_ref_comp|MaxEntScan_alt|MaxEntScan_diff|MaxEntScan_ref|GERP++_NR|GERP++_RS|REVEL_rankscore|REVEL_score|phastCons100way_vertebrate|phyloP100way_vertebrate|rs_dbSNP150|LoFtool|pLI_gene_value|SpliceAI_pred_DP_AG|SpliceAI_pred_DP_AL|SpliceAI_pred_DP_DG|SpliceAI_pred_DP_DL|SpliceAI_pred_DS_AG|SpliceAI_pred_DS_AL|SpliceAI_pred_DS_DG|SpliceAI_pred_DS_DL|SpliceAI_pred_SYMBOL|genomic_superdups_frac_match'\n",
    "csq_name_split=csq_name.split('|')\n",
    "\n",
    "## get two fields' index\n",
    "con_idx = csq_name_split.index('Consequence') \n",
    "CANONICALidx = csq_name_split.index('CANONICAL')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"4. for CSQ: give Consequence order \"\"\"\n",
    "\n",
    "with open('variant_consequences.txt','r') as f:            # get the 'variant_consequences order' lst: descending severity\n",
    "    order_conseq=[line.rstrip(\"\\n\") for line in f]         # remove the \\n at right side of each line; lst\n",
    "\n",
    "# give consequence order number\n",
    "orderdic = {v:i for i,v in enumerate(order_conseq)}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"5.for CSQ, function: split transcripts into fields, and clean Consequence cell\"\"\"\n",
    "\n",
    "def split_CSQ_fields(trans_csq,order_conseq,con_idx):   \n",
    "\n",
    "    for i in range(len(trans_csq)):\n",
    "            split_trans = trans_csq[i].split('|')         # split each trans into fields as str\n",
    "                                                                                                             \n",
    "            cons_sub = split_trans[con_idx].split('&')    # in each trans: split `Consequence` cell inside\n",
    "            cons_sub.sort(key=order_conseq.index)         # in each trans: order `Consequence` cell inside\n",
    "            split_trans[con_idx] = cons_sub[0]            # in each trans: only keep the most severe `Consequence` in this cell\n",
    "            \n",
    "            trans_csq[i] = split_trans                    # 2D lst\n",
    "    return trans_csq\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"6.for CSQ, function: pick one trans of each variant: `Conseqence` is most severe and `Canonical`==Yes \"\"\"\n",
    "\n",
    "def pick_CSQ_trans_final(trans_csq,orderdic,con_idx,CANONICALidx):\n",
    "\n",
    "    ### order and pick one trans of each variant\n",
    "    trans_csq.sort(key=lambda x:orderdic[x[con_idx]])     # in each variant: order trans by `Consequence` cell\n",
    "\n",
    "    found = -1\n",
    "    for i in range(len(trans_csq)):                       \n",
    "        if(trans_csq[i][CANONICALidx] == \"YES\" and trans_csq[i][con_idx] == trans_csq[0][con_idx]):         \n",
    "            found = i                                 \n",
    "            break\n",
    "    if(not found==-1):\n",
    "        picked_trans=(trans_csq[found] )                  # in each variant: if any trans `Cons` is the most sever and `Canonical`=YES, pick this trans\n",
    "    else:                                                 # if no any `Canonical`= yes, pick the fist most sever trans\n",
    "        picked_trans=(trans_csq[0])\n",
    "\n",
    "    return picked_trans\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\" 7. main script: extract all features outside and inside CSQ\"\"\"\n",
    "\n",
    "in_file = '/Users/nancy/Desktop/RS_projects/data/02_data_analysis/00_original_data/clinvar_221113.vcf.gz'\n",
    "out_file = '/Users/nancy/Desktop/RS_projects/data/03_ML/01_featureV1/01_train/modelV2/01_ALLfeatures_extracted_addCLNSIG.csv'\n",
    "\n",
    " \n",
    "head_line = float('inf')                                            #  as infinity integer\n",
    "\n",
    "with gzip.open(in_file) as in_f:                                    # read compressed file\n",
    "    with open(out_file,'w') as out_f:                               # write new file.csv\n",
    "        \n",
    "        ### write the head of features outside CSQ and CSQ\n",
    "        out_f.write(';'.join(features) +';')                        \n",
    "        out_f.write(';'.join(csq_name_split)+'\\n')                  \n",
    "\n",
    "\n",
    "        for i,line in enumerate(in_f,0):                           \n",
    "            content = line.decode('utf8').rstrip('\\n')             \n",
    "            if content.startswith('#CHROM'):                        # find first line of rec \n",
    "                rec_head=content.split('\\t')                        # rec head line: split str to list\n",
    "                INFOidx = rec_head.index('INFO')                    # get INFO idx\n",
    "                head_line = i\n",
    "\n",
    "\n",
    "            if i > head_line:\n",
    "                rec_lst=content.split('\\t')                                                  # lst: get each variant line , split str\n",
    "\n",
    "                ### extract features outside CSQ\n",
    "                info = rec_lst[INFOidx].split(';')                                           # lst: get each INFO line               \n",
    "                feature_line = extract_feature(features,feature_value_start,info)            # get each feature line\n",
    "\n",
    "                ### extract CSQ\n",
    "                trans_csq = rec_lst[INFOidx].split('CSQ=')[-1].split(\";\")[0].split(\",\")       # in each variant: get csq's transcripts , lst\n",
    "                trans_csq = split_CSQ_fields(trans_csq,order_conseq,con_idx)                  # in each variant: splite each transcript into fields, and ordered by `Consequence``\n",
    "                picked_trans = pick_CSQ_trans_final(trans_csq,orderdic,con_idx,CANONICALidx)  # in each variant: pick one transcript which `Conseqence` is most severe and `Canonical`==Yes \"\"\"\n",
    "                \n",
    "                ### write down\n",
    "                out_f.write(';'.join(feature_line) +';')           \n",
    "                out_f.write(';'.join(picked_trans) +'\\n')           # why ';' instead of ','--- since CLNSIG has'Likely_pathogenic,_low_penetrance' with ',' inside to mess the formate CSV\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data preprocessing pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/q0/2p8kfxsn7s73349btf1_g07h0000gp/T/ipykernel_4968/3969921985.py:1: DtypeWarning: Columns (32,49,50,51,54,56,58,92) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('/Users/nancy/Desktop/RS_projects/data/03_ML/01_train_model/01_featureV1/01_ALLfeatures_extracted_addCLNSIG.csv',sep=';',index_col=False)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CADD</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>Frq</th>\n",
       "      <th>GNOMADAF</th>\n",
       "      <th>GNOMADAF_popmax</th>\n",
       "      <th>Hom</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>...</th>\n",
       "      <th>SpliceAI_pred_DP_AG</th>\n",
       "      <th>SpliceAI_pred_DP_AL</th>\n",
       "      <th>SpliceAI_pred_DP_DG</th>\n",
       "      <th>SpliceAI_pred_DP_DL</th>\n",
       "      <th>SpliceAI_pred_DS_AG</th>\n",
       "      <th>SpliceAI_pred_DS_AL</th>\n",
       "      <th>SpliceAI_pred_DS_DG</th>\n",
       "      <th>SpliceAI_pred_DS_DL</th>\n",
       "      <th>SpliceAI_pred_SYMBOL</th>\n",
       "      <th>genomic_superdups_frac_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>26.600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-30.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>-30.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>13.420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>-34.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>31.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.000414</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>41.0</td>\n",
       "      <td>-47.0</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>28.200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>-47.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>11.380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-35.0</td>\n",
       "      <td>-50.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468907</th>\n",
       "      <td>Benign</td>\n",
       "      <td>7.587</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468908</th>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>23.800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468909</th>\n",
       "      <td>Benign</td>\n",
       "      <td>11.200</td>\n",
       "      <td>0.00614</td>\n",
       "      <td>0.00589</td>\n",
       "      <td>0.00243</td>\n",
       "      <td>0.00671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468910</th>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>21.600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468911</th>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>5.338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.02530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1468912 rows × 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         CLNSIG    CADD   AF_ESP  AF_EXAC   AF_TGP      Frq  \\\n",
       "0        Uncertain_significance  26.600      NaN      NaN      NaN      NaN   \n",
       "1                 Likely_benign  13.420      NaN      NaN      NaN      NaN   \n",
       "2                 Likely_benign  31.000      NaN      NaN      NaN  0.00056   \n",
       "3        Uncertain_significance  28.200      NaN      NaN      NaN      NaN   \n",
       "4                 Likely_benign  11.380      NaN      NaN      NaN      NaN   \n",
       "...                         ...     ...      ...      ...      ...      ...   \n",
       "1468907                  Benign   7.587      NaN      NaN      NaN  0.00070   \n",
       "1468908  Uncertain_significance  23.800      NaN      NaN      NaN      NaN   \n",
       "1468909                  Benign  11.200  0.00614  0.00589  0.00243  0.00671   \n",
       "1468910  Uncertain_significance  21.600      NaN      NaN      NaN      NaN   \n",
       "1468911           Likely_benign   5.338      NaN  0.00000      NaN  0.02530   \n",
       "\n",
       "         GNOMADAF  GNOMADAF_popmax   Hom  ORIGIN  ...  SpliceAI_pred_DP_AG  \\\n",
       "0             NaN              NaN   NaN     1.0  ...                -30.0   \n",
       "1             NaN              NaN   NaN     1.0  ...                  7.0   \n",
       "2        0.000414         0.000195   NaN     1.0  ...                 41.0   \n",
       "3             NaN              NaN   NaN     1.0  ...                 10.0   \n",
       "4             NaN              NaN   NaN     1.0  ...                -35.0   \n",
       "...           ...              ...   ...     ...  ...                  ...   \n",
       "1468907       NaN              NaN   5.0     1.0  ...                 25.0   \n",
       "1468908       NaN              NaN   NaN    32.0  ...                  NaN   \n",
       "1468909       NaN              NaN  48.0     1.0  ...                  NaN   \n",
       "1468910       NaN              NaN   NaN     1.0  ...                  NaN   \n",
       "1468911       NaN              NaN  29.0     1.0  ...                  NaN   \n",
       "\n",
       "         SpliceAI_pred_DP_AL SpliceAI_pred_DP_DG SpliceAI_pred_DP_DL  \\\n",
       "0                       11.0                10.0               -30.0   \n",
       "1                      -34.0                 6.0                42.0   \n",
       "2                      -47.0                -7.0                44.0   \n",
       "3                      -47.0                34.0               -14.0   \n",
       "4                      -50.0                12.0                24.0   \n",
       "...                      ...                 ...                 ...   \n",
       "1468907                 35.0                 0.0                16.0   \n",
       "1468908                  NaN                 NaN                 NaN   \n",
       "1468909                  NaN                 NaN                 NaN   \n",
       "1468910                  NaN                 NaN                 NaN   \n",
       "1468911                  NaN                 NaN                 NaN   \n",
       "\n",
       "        SpliceAI_pred_DS_AG SpliceAI_pred_DS_AL SpliceAI_pred_DS_DG  \\\n",
       "0                      0.01                0.00                0.09   \n",
       "1                      0.00                0.02                0.03   \n",
       "2                      0.00                0.01                0.06   \n",
       "3                      0.00                0.00                0.03   \n",
       "4                      0.00                0.00                0.01   \n",
       "...                     ...                 ...                 ...   \n",
       "1468907                0.00                0.00                0.00   \n",
       "1468908                 NaN                 NaN                 NaN   \n",
       "1468909                 NaN                 NaN                 NaN   \n",
       "1468910                 NaN                 NaN                 NaN   \n",
       "1468911                 NaN                 NaN                 NaN   \n",
       "\n",
       "        SpliceAI_pred_DS_DL SpliceAI_pred_SYMBOL genomic_superdups_frac_match  \n",
       "0                      0.00               SAMD11                          NaN  \n",
       "1                      0.00               SAMD11                          NaN  \n",
       "2                      0.00               SAMD11                          NaN  \n",
       "3                      0.02               SAMD11                          NaN  \n",
       "4                      0.00               SAMD11                          NaN  \n",
       "...                     ...                  ...                          ...  \n",
       "1468907                0.00                USP9Y                          NaN  \n",
       "1468908                 NaN                  NaN                          NaN  \n",
       "1468909                 NaN                  NaN                          NaN  \n",
       "1468910                 NaN                  NaN                          NaN  \n",
       "1468911                 NaN                  NaN                          NaN  \n",
       "\n",
       "[1468912 rows x 93 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/Users/nancy/Desktop/RS_projects/data/03_ML/01_train_model/01_featureV1/01_ALLfeatures_extracted_addCLNSIG.csv',sep=';',index_col=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" 1. choose featuresV1\"\"\"\n",
    "\n",
    "featureV1 = ['CLNSIG','CADD', 'AF_TGP', 'Frq', 'IMPACT','GNOMADAF_popmax', 'Hom', 'ORIGIN',\n",
    "       'SPIDEX', 'SWEGENAF', 'Consequence', 'BIOTYPE', 'SIFT',\n",
    "       'PolyPhen', 'MES-SWA_acceptor_alt', 'MES-SWA_acceptor_diff',\n",
    "       'MES-SWA_donor_alt', 'MES-SWA_donor_diff', 'MaxEntScan_alt',\n",
    "       'MaxEntScan_diff', 'GERP++_RS', 'REVEL_score',\n",
    "       'phastCons100way_vertebrate', 'phyloP100way_vertebrate', 'LoFtool',\n",
    "       'pLI_gene_value', 'SpliceAI_pred_DS_AG', 'SpliceAI_pred_DS_AL',\n",
    "       'SpliceAI_pred_DS_DG', 'SpliceAI_pred_DS_DL']         # keep CLNSIG,\n",
    "\n",
    "df_1 = df[featureV1]\n",
    "\n",
    "\n",
    "\n",
    "\"\"\" 2. only remove missing target \"\"\"\n",
    "# for some features with fewer missing\n",
    "# drop_nan_features = ['CLNSIG','CADD','Consequence','IMPACT','BIOTYPE','MES-SWA_acceptor_alt','MES-SWA_acceptor_diff',\n",
    "# 'MES-SWA_donor_alt','MES-SWA_donor_diff','pLI_gene_value']\n",
    "\n",
    "drop_nan_features = ['CLNSIG']\n",
    "df_1 = df_1.dropna(subset=drop_nan_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(featureV1)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" 3. deal with target feature \"\"\"\n",
    "# \"\"\"  CLNSIG: tansfer to 4 classes__ benign,pathogenic,uncertain,others\n",
    "# NOTE:\n",
    "# (only count name before |)   explanation-- https://www.ncbi.nlm.nih.gov/clinvar/docs/clinsig/#clinsig_scv\n",
    "\n",
    "# 1. pathogenic: Pathogenic, Pathogenic/Likely_pathogenic, Likely_pathogenic\n",
    "# 2. benign: Likely_benign, Benign, Benign/Likely_benign\n",
    "# 3. uncertain: Uncertain_significance, \n",
    "# 4. others: Uncertain_risk_allele, risk_factor, protective, other, not_provided, Likely_risk_allele, drug_response, Conflicting_interpretations_of_pathogenicity, confers_sensitivity, associatio, Affects\n",
    "\n",
    "df_1 = df_1.copy()\n",
    "df_1['CLNSIG'] = df_1['CLNSIG'].replace(['Uncertain_significance|_risk_factor', 'Uncertain_significance|_other', 'Uncertain_significance|_drug_response','Uncertain_significance|_association','Uncertain_significance|_Affects','Uncertain_significance'], 'uncertain')\n",
    "df_1['CLNSIG'] = df_1['CLNSIG'].replace(['Pathogenic|_risk_factor','Pathogenic|_protective','Pathogenic|_other','Pathogenic|_drug_response|_other','Pathogenic|_drug_response','Pathogenic|_confers_sensitivity','Pathogenic|_association','Pathogenic|_Affects','Pathogenic/Likely_risk_allele','Pathogenic/Likely_pathogenic|_risk_factor','Pathogenic/Likely_pathogenic|_other','Pathogenic/Likely_pathogenic|_drug_response','Pathogenic/Likely_pathogenic','Pathogenic','Likely_pathogenic|_risk_factor','Likely_pathogenic|_other','Likely_pathogenic|_drug_response','Likely_pathogenic|_association','Likely_pathogenic|_Affects','Likely_pathogenic,_low_penetrance','Likely_pathogenic'], 'pathogenic')\n",
    "df_1['CLNSIG'] = df_1['CLNSIG'].replace(['Likely_benign|_risk_factor','Likely_benign|_other','Likely_benign|_drug_response|_other','Likely_benign','Benign|_risk_factor','Benign|_protective','Benign|_other','Benign|_drug_response','Benign|_confers_sensitivity','Benign|_association|_confers_sensitivity','Benign|_association','Benign/Likely_benign|_risk_factor','Benign/Likely_benign|_other|_risk_factor','Benign/Likely_benign|_other','Benign/Likely_benign|_drug_response|_other','Benign/Likely_benign|_drug_response','Benign/Likely_benign|_association','Benign/Likely_benign','Benign'], 'benign')\n",
    "df_1['CLNSIG'] = df_1['CLNSIG'].replace(['Uncertain_risk_allele|_risk_factor','Uncertain_risk_allele','risk_factor','protective','protective|_risk_factor','other', 'not_provided','Likely_risk_allele','drug_response|_risk_factor','drug_response|_other','drug_response','Conflicting_interpretations_of_pathogenicity|_risk_factor','Conflicting_interpretations_of_pathogenicity|_protective','Conflicting_interpretations_of_pathogenicity|_other|_risk_factor','Conflicting_interpretations_of_pathogenicity|_other','Conflicting_interpretations_of_pathogenicity|_drug_response|_other','Conflicting_interpretations_of_pathogenicity|_drug_response','Conflicting_interpretations_of_pathogenicity|_association|_risk_factor','Conflicting_interpretations_of_pathogenicity|_association','Conflicting_interpretations_of_pathogenicity','confers_sensitivity','association|_risk_factor','association_not_found','association','Affects|_risk_factor','Affects|_association','Affects'], 'others')\n",
    "\n",
    "# only keep benign & pathogenic\n",
    "df_1 = df_1.loc[df_1['CLNSIG'].isin(['benign','pathogenic'])]\n",
    "\n",
    "# label encoder\n",
    "map_clnsig={'pathogenic':1,'benign':0}\n",
    "df_1['CLNSIG']=df_1['CLNSIG'].map(lambda s: map_clnsig.get(s) if s in map_clnsig else s)\n",
    "# df_1[\"CLNSIG\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CADD</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>Frq</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>GNOMADAF_popmax</th>\n",
       "      <th>Hom</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>SPIDEX</th>\n",
       "      <th>SWEGENAF</th>\n",
       "      <th>...</th>\n",
       "      <th>GERP++_RS</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>phastCons100way_vertebrate</th>\n",
       "      <th>phyloP100way_vertebrate</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>pLI_gene_value</th>\n",
       "      <th>SpliceAI_pred_DS_AG</th>\n",
       "      <th>SpliceAI_pred_DS_AL</th>\n",
       "      <th>SpliceAI_pred_DS_DG</th>\n",
       "      <th>SpliceAI_pred_DS_DL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>13.420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>31.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4.52</td>\n",
       "      <td>0.429</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>11.380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>9.519</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>5.182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468905</th>\n",
       "      <td>0</td>\n",
       "      <td>4.853</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.18160</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.274000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468906</th>\n",
       "      <td>1</td>\n",
       "      <td>26.500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00014</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468907</th>\n",
       "      <td>0</td>\n",
       "      <td>7.587</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00070</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468909</th>\n",
       "      <td>0</td>\n",
       "      <td>11.200</td>\n",
       "      <td>0.00243</td>\n",
       "      <td>0.00671</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.796</td>\n",
       "      <td>0.001927</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468911</th>\n",
       "      <td>0</td>\n",
       "      <td>5.338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.02530</td>\n",
       "      <td>LOW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.157000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>790852 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CLNSIG    CADD   AF_TGP      Frq    IMPACT  GNOMADAF_popmax   Hom  \\\n",
       "1             0  13.420      NaN      NaN       LOW              NaN   NaN   \n",
       "2             0  31.000      NaN  0.00056  MODERATE         0.000195   NaN   \n",
       "4             0  11.380      NaN      NaN       LOW              NaN   NaN   \n",
       "6             0   9.519      NaN      NaN       LOW              NaN   NaN   \n",
       "8             0   5.182      NaN      NaN       LOW              NaN   NaN   \n",
       "...         ...     ...      ...      ...       ...              ...   ...   \n",
       "1468905       0   4.853      NaN  0.18160       LOW              NaN   NaN   \n",
       "1468906       1  26.500      NaN  0.00014      HIGH              NaN   1.0   \n",
       "1468907       0   7.587      NaN  0.00070       LOW              NaN   5.0   \n",
       "1468909       0  11.200  0.00243  0.00671       LOW              NaN  48.0   \n",
       "1468911       0   5.338      NaN  0.02530       LOW              NaN  29.0   \n",
       "\n",
       "         ORIGIN  SPIDEX  SWEGENAF  ... GERP++_RS REVEL_score  \\\n",
       "1           1.0  -1.030       NaN  ...       NaN         NaN   \n",
       "2           1.0  -0.910       NaN  ...      4.52       0.429   \n",
       "4           1.0   0.992       NaN  ...       NaN         NaN   \n",
       "6           1.0  -1.049       NaN  ...       NaN         NaN   \n",
       "8           1.0  -1.021       NaN  ...       NaN         NaN   \n",
       "...         ...     ...       ...  ...       ...         ...   \n",
       "1468905     1.0     NaN  0.274000  ...       NaN         NaN   \n",
       "1468906     1.0     NaN       NaN  ...       NaN         NaN   \n",
       "1468907     1.0  -0.088       NaN  ...       NaN         NaN   \n",
       "1468909     1.0  -1.796  0.001927  ...       NaN         NaN   \n",
       "1468911     1.0     NaN  0.157000  ...       NaN         NaN   \n",
       "\n",
       "        phastCons100way_vertebrate phyloP100way_vertebrate  LoFtool  \\\n",
       "1                              NaN                     NaN      NaN   \n",
       "2                              1.0                   5.538      NaN   \n",
       "4                              NaN                     NaN      NaN   \n",
       "6                              NaN                     NaN      NaN   \n",
       "8                              NaN                     NaN      NaN   \n",
       "...                            ...                     ...      ...   \n",
       "1468905                        NaN                     NaN      NaN   \n",
       "1468906                        NaN                     NaN      NaN   \n",
       "1468907                        NaN                     NaN      NaN   \n",
       "1468909                        NaN                     NaN      NaN   \n",
       "1468911                        NaN                     NaN      NaN   \n",
       "\n",
       "         pLI_gene_value  SpliceAI_pred_DS_AG  SpliceAI_pred_DS_AL  \\\n",
       "1                  0.00                 0.00                 0.02   \n",
       "2                  0.00                 0.00                 0.01   \n",
       "4                  0.00                 0.00                 0.00   \n",
       "6                  0.00                 0.00                 0.00   \n",
       "8                  0.01                 0.03                 0.00   \n",
       "...                 ...                  ...                  ...   \n",
       "1468905            0.31                 0.00                 0.00   \n",
       "1468906            0.00                 0.00                 0.00   \n",
       "1468907            0.00                 0.00                 0.00   \n",
       "1468909            0.10                  NaN                  NaN   \n",
       "1468911            0.10                  NaN                  NaN   \n",
       "\n",
       "         SpliceAI_pred_DS_DG  SpliceAI_pred_DS_DL  \n",
       "1                       0.03                 0.00  \n",
       "2                       0.06                 0.00  \n",
       "4                       0.01                 0.00  \n",
       "6                       0.05                 0.17  \n",
       "8                       0.00                 0.00  \n",
       "...                      ...                  ...  \n",
       "1468905                 0.00                 0.00  \n",
       "1468906                 0.00                 1.00  \n",
       "1468907                 0.00                 0.00  \n",
       "1468909                  NaN                  NaN  \n",
       "1468911                  NaN                  NaN  \n",
       "\n",
       "[790852 rows x 30 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" 4. split data into train and test\"\"\" \n",
    "\n",
    "\"\"\" try to train twice, since imbalance classes\n",
    "\n",
    "4.1 make balanced classes: undersampling the benign samples to make two sample uneven\"\"\"\n",
    "\n",
    "# Undersampling: balance uneven benign and pathogenic quantity\n",
    "# benign = df_1[df_1[\"CLNSIG\"]==0]\n",
    "# pathogenic = df_1[df_1[\"CLNSIG\"]==1]\n",
    "# benign = benign.sample(len(pathogenic),random_state=42)   # randomly extract benign same quantity with patho\n",
    "# data = pd.concat( (benign,pathogenic),axis=0)\n",
    "\n",
    "# # split\n",
    "# from sklearn.model_selection import train_test_split\n",
    "\n",
    "# X, y = data.iloc[:, 1:], data.iloc[:, 0]\n",
    "# X_train, X_test, y_train, y_test =\\\n",
    "#     train_test_split(X, y, \n",
    "#                     test_size=0.3,\n",
    "#                     random_state=42,\n",
    "#                     stratify=y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"4.2  keep imbalance classes \"\"\"\n",
    "\n",
    "# split\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "X, y = df_1.iloc[:, 1:], df_1.iloc[:, 0]\n",
    "X_train, X_test, y_train, y_test =\\\n",
    "    train_test_split(X, y, \n",
    "                    test_size=0.3,\n",
    "                    random_state=42,\n",
    "                    stratify=y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_consequences_order = './variant_consequences.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old version: include remove missing data features\n",
    "# \n",
    "# \"\"\" 5. features engineering: transformer\"\"\"\n",
    "# ## only for training data, no care about target 'CLNSIG'\n",
    "\n",
    "# from sklearn.compose import make_column_transformer\n",
    "# from sklearn.impute import MissingIndicator, SimpleImputer\n",
    "# from sklearn.preprocessing import OrdinalEncoder, StandardScaler\n",
    "# from category_encoders import BinaryEncoder\n",
    "\n",
    "# ## numerical feature group 1 : replace missing with mean;  scaling/standardization\n",
    "\n",
    "# numeric_feature_1 = ['MaxEntScan_alt','MaxEntScan_diff','GERP++_RS',\n",
    "#     'phastCons100way_vertebrate','phyloP100way_vertebrate']\n",
    "\n",
    "\n",
    "\n",
    "# ## numerical feature group 2 : no missing left or/ replace missing with 0;  scaling/standardization\n",
    "# numeric_feature_2 = ['CADD','AF_TGP', 'Frq', 'GNOMADAF_popmax', 'Hom',\n",
    "#        'SWEGENAF','MES-SWA_acceptor_alt','MES-SWA_acceptor_diff','MES-SWA_donor_alt',\n",
    "#        'MES-SWA_donor_diff','LoFtool','pLI_gene_value','SpliceAI_pred_DS_AL',\n",
    "#        'SpliceAI_pred_DS_DG','SpliceAI_pred_DS_DL']\n",
    "\n",
    "\n",
    "\n",
    "# ## numerical feature group 3: add missing indicator; replace missing with 0 and ; scaling/standardization\n",
    "# numeric_feature_3 = ['SPIDEX','REVEL_score','SpliceAI_pred_DS_AG']\n",
    "\n",
    "\n",
    "\n",
    "# ## categorical ordinal feature group 4: add missing indicator; no missing left or/ replace missing with 0; OrdinalEncoder\n",
    "# categori_oridinal_feature_4 =['Consequence','IMPACT','SIFT','PolyPhen']\n",
    "\n",
    "# # ordinal features' order: left to right will be from 0 to length-1\n",
    "# with open(variant_consequences_order,'r') as f:             # get the 'variant_consequences order' lst: descending severity\n",
    "#     order_conseq=[line.rstrip(\"\\n\") for line in f] \n",
    "#     order_conseq.reverse()\n",
    "\n",
    "# ordinal_features = [\n",
    "#     'Consequence',\n",
    "#     'IMPACT',\n",
    "#     'SIFT',\n",
    "#     'PolyPhen']\n",
    "# # 0 is the missing value will be replaced by\n",
    "# ordinal_ordering = [\n",
    "#     order_conseq,\n",
    "#     ['MODIFIER','LOW','MODERATE','HIGH'],\n",
    "#     [0,'tolerated_low_confidence','tolerated','deleterious_low_confidence','deleterious'],\n",
    "#     [0,'unknown','benign','possibly_damaging','probably_damaging']\n",
    "#     ]\n",
    "\n",
    "\n",
    "# ## categorical nominal feature group 5: no missing left or/ replace missing with 0; BinaryEncoder\n",
    "# categori_nominal_feature_5 = ['ORIGIN','BIOTYPE']  #,'BIOTYPE'\n",
    "\n",
    "\n",
    "# from sklearn.pipeline import Pipeline, make_pipeline\n",
    "# from sklearn.compose import ColumnTransformer, make_column_transformer\n",
    "\n",
    "# numeric_feature_1_transformer =make_pipeline(\n",
    "#     SimpleImputer(strategy='mean'),\n",
    "#     StandardScaler()\n",
    "# )\n",
    "\n",
    "# numeric_feature_2_transformer =make_pipeline(\n",
    "#     SimpleImputer(strategy='constant',fill_value=0),\n",
    "#     StandardScaler()\n",
    "# )\n",
    "\n",
    "# numeric_feature_3_transformer =make_pipeline(\n",
    "#     SimpleImputer(strategy='constant',fill_value=0),\n",
    "#     StandardScaler()\n",
    "# )\n",
    "\n",
    "# categori_oridinal_feature_4_transformer =make_pipeline(\n",
    "#     SimpleImputer(strategy='constant',fill_value=0), \n",
    "#     OrdinalEncoder(categories=ordinal_ordering)\n",
    "# )\n",
    "\n",
    "# categori_oridinal_feature_5_transformer =make_pipeline(\n",
    "#     SimpleImputer(strategy='constant',fill_value=0),\n",
    "#     BinaryEncoder()                                # add new columns, throw the original columns automaticly, \n",
    "# )\n",
    "\n",
    "\n",
    "# preprocessor = make_column_transformer(\n",
    "#     (numeric_feature_1_transformer, numeric_feature_1),\n",
    "#     (numeric_feature_2_transformer, numeric_feature_2 ),\n",
    "#     (MissingIndicator(), numeric_feature_3 + categori_oridinal_feature_4),   # add new missing indicators for feature group 3-4 without originial columns,  here: add 5 indicator columns in total\n",
    "#     (numeric_feature_3_transformer, numeric_feature_3),  \n",
    "#     (categori_oridinal_feature_4_transformer, categori_oridinal_feature_4),\n",
    "#     (categori_oridinal_feature_5_transformer, categori_nominal_feature_5),   # here: 'ORIGIN': 6 new, 'BIOTYPE': 3 new\n",
    "#     remainder='passthrough'\n",
    "\n",
    "# )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" 5. features engineering: transformer\"\"\"\n",
    "## only for training data, no care about target 'CLNSIG', total 29 features\n",
    "\n",
    "from sklearn.compose import make_column_transformer\n",
    "from sklearn.impute import MissingIndicator, SimpleImputer\n",
    "from sklearn.preprocessing import OrdinalEncoder, StandardScaler\n",
    "from category_encoders import BinaryEncoder\n",
    "from sklearn.pipeline import Pipeline, make_pipeline\n",
    "from sklearn.compose import ColumnTransformer, make_column_transformer\n",
    "\n",
    "## numerical feature group 1 : replace missing with mean;  scaling/standardization\n",
    "\n",
    "numeric_feature_1 = ['MES-SWA_acceptor_alt','MES-SWA_acceptor_diff','MES-SWA_donor_alt','MES-SWA_donor_diff','MaxEntScan_alt','MaxEntScan_diff','GERP++_RS',\n",
    "    'phastCons100way_vertebrate','phyloP100way_vertebrate']\n",
    "\n",
    "\n",
    "\n",
    "## numerical feature group 2 : replace missing with 0;  scaling/standardization\n",
    "numeric_feature_2 = ['CADD','AF_TGP', 'Frq', 'GNOMADAF_popmax', 'Hom',\n",
    "       'SWEGENAF','LoFtool','pLI_gene_value','SpliceAI_pred_DS_AL',\n",
    "       'SpliceAI_pred_DS_DG','SpliceAI_pred_DS_DL']\n",
    "\n",
    "\n",
    "\n",
    "## numerical feature group 3: add missing indicator; replace missing with 0 and ; scaling/standardization\n",
    "numeric_feature_3 = ['SPIDEX','REVEL_score','SpliceAI_pred_DS_AG']\n",
    "\n",
    "\n",
    "\n",
    "## categorical ordinal feature group 4: add missing indicator for ['SIFT','PolyPhen']; replace missing with 'missing_value' string; OrdinalEncoder\n",
    "categori_oridinal_feature_4 =['Consequence','IMPACT','SIFT','PolyPhen']\n",
    "\n",
    "# ordinal features' order: left to right will be from 0 to length-1\n",
    "with open(variant_consequences_order,'r') as f:             # get the 'variant_consequences order' lst: descending severity\n",
    "    order_conseq=[line.rstrip(\"\\n\") for line in f] \n",
    "    order_conseq.reverse()\n",
    "\n",
    "ordinal_features = [\n",
    "    'Consequence',\n",
    "    'IMPACT',\n",
    "    'SIFT',\n",
    "    'PolyPhen']\n",
    "# 0 is the missing value will be replaced by\n",
    "ordinal_ordering = [\n",
    "    order_conseq,\n",
    "    ['MODIFIER','LOW','MODERATE','HIGH'],\n",
    "    ['tolerated_low_confidence','tolerated','deleterious_low_confidence','deleterious'],\n",
    "    ['unknown','benign','possibly_damaging','probably_damaging']\n",
    "    ]\n",
    "\n",
    "\n",
    "## categorical nominal feature group 5: replace missing with 0 in ['ORIGIN'], with ''missing_value' in ['BIOTYPE'];  BinaryEncoder\n",
    "categori_nominal_feature_5 = ['ORIGIN','BIOTYPE']  #,'BIOTYPE'\n",
    "\n",
    "\n",
    "\n",
    "numeric_feature_1_transformer =make_pipeline(\n",
    "    SimpleImputer(strategy='mean'),\n",
    "    StandardScaler()\n",
    ")\n",
    "\n",
    "numeric_feature_2_transformer =make_pipeline(\n",
    "    SimpleImputer(strategy='constant',fill_value=0),\n",
    "    StandardScaler()\n",
    ")\n",
    "\n",
    "numeric_feature_3_transformer =make_pipeline(\n",
    "    SimpleImputer(strategy='constant',fill_value=0),\n",
    "    StandardScaler()\n",
    ")\n",
    "\n",
    "categori_oridinal_feature_4_transformer =make_pipeline(\n",
    "    SimpleImputer(strategy='constant',fill_value=None),  # If None, fill_value will be 0 when imputing numerical data and “missing_value” for strings\n",
    "    OrdinalEncoder(categories=ordinal_ordering, handle_unknown='use_encoded_value', unknown_value = -1)       # the unknown 'missing_value' string will be enocoded as '-1'\n",
    ")\n",
    "\n",
    "categori_oridinal_feature_5_transformer =make_pipeline(\n",
    "    SimpleImputer(strategy='constant',fill_value=None),  \n",
    "    BinaryEncoder()                                # add new columns, throw the original columns automaticly, \n",
    ")\n",
    "\n",
    "\n",
    "preprocessor = make_column_transformer(\n",
    "    (numeric_feature_1_transformer, numeric_feature_1),\n",
    "    (numeric_feature_2_transformer, numeric_feature_2 ),\n",
    "    (MissingIndicator(), numeric_feature_3 + categori_oridinal_feature_4[2:]),   # add new missing indicators for feature3 + ['SIFT','PolyPhen'] without originial columns,  here: total add 5 indicator columns\n",
    "    (numeric_feature_3_transformer, numeric_feature_3),  \n",
    "    (categori_oridinal_feature_4_transformer, categori_oridinal_feature_4),\n",
    "    (categori_oridinal_feature_5_transformer, categori_nominal_feature_5),   # here: 'ORIGIN': 6 new, 'BIOTYPE': 3 new\n",
    "    # remainder='passthrough'                                                   # \n",
    "\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-431e2d62-9134-43e3-93d4-90b68688e427 {color: black;background-color: white;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 pre{padding: 0;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-toggleable {background-color: white;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-estimator:hover {background-color: #d4ebff;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 2em;bottom: 0;left: 50%;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-item {z-index: 1;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-parallel::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 2em;bottom: 0;left: 50%;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-parallel-item {display: flex;flex-direction: column;position: relative;background-color: white;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-parallel-item:only-child::after {width: 0;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;position: relative;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-label label {font-family: monospace;font-weight: bold;background-color: white;display: inline-block;line-height: 1.2em;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-label-container {position: relative;z-index: 2;text-align: center;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-431e2d62-9134-43e3-93d4-90b68688e427 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-431e2d62-9134-43e3-93d4-90b68688e427\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>ColumnTransformer(transformers=[(&#x27;pipeline-1&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;simpleimputer&#x27;,\n",
       "                                                  SimpleImputer()),\n",
       "                                                 (&#x27;standardscaler&#x27;,\n",
       "                                                  StandardScaler())]),\n",
       "                                 [&#x27;MES-SWA_acceptor_alt&#x27;,\n",
       "                                  &#x27;MES-SWA_acceptor_diff&#x27;, &#x27;MES-SWA_donor_alt&#x27;,\n",
       "                                  &#x27;MES-SWA_donor_diff&#x27;, &#x27;MaxEntScan_alt&#x27;,\n",
       "                                  &#x27;MaxEntScan_diff&#x27;, &#x27;GERP++_RS&#x27;,\n",
       "                                  &#x27;phastCons100way_vertebrate&#x27;,\n",
       "                                  &#x27;phyloP100way_vertebrate&#x27;]),\n",
       "                                (&#x27;pipeline-2&#x27;,\n",
       "                                 Pipeline(...\n",
       "                                                                              &#x27;tolerated&#x27;,\n",
       "                                                                              &#x27;deleterious_low_confidence&#x27;,\n",
       "                                                                              &#x27;deleterious&#x27;],\n",
       "                                                                             [&#x27;unknown&#x27;,\n",
       "                                                                              &#x27;benign&#x27;,\n",
       "                                                                              &#x27;possibly_damaging&#x27;,\n",
       "                                                                              &#x27;probably_damaging&#x27;]],\n",
       "                                                                 handle_unknown=&#x27;use_encoded_value&#x27;,\n",
       "                                                                 unknown_value=-1))]),\n",
       "                                 [&#x27;Consequence&#x27;, &#x27;IMPACT&#x27;, &#x27;SIFT&#x27;, &#x27;PolyPhen&#x27;]),\n",
       "                                (&#x27;pipeline-5&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;simpleimputer&#x27;,\n",
       "                                                  SimpleImputer(strategy=&#x27;constant&#x27;)),\n",
       "                                                 (&#x27;binaryencoder&#x27;,\n",
       "                                                  BinaryEncoder())]),\n",
       "                                 [&#x27;ORIGIN&#x27;, &#x27;BIOTYPE&#x27;])])</pre><b>Please rerun this cell to show the HTML repr or trust the notebook.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"df8d1d6c-2262-4fa1-bfe7-2e7740cb53a6\" type=\"checkbox\" ><label for=\"df8d1d6c-2262-4fa1-bfe7-2e7740cb53a6\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">ColumnTransformer</label><div class=\"sk-toggleable__content\"><pre>ColumnTransformer(transformers=[(&#x27;pipeline-1&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;simpleimputer&#x27;,\n",
       "                                                  SimpleImputer()),\n",
       "                                                 (&#x27;standardscaler&#x27;,\n",
       "                                                  StandardScaler())]),\n",
       "                                 [&#x27;MES-SWA_acceptor_alt&#x27;,\n",
       "                                  &#x27;MES-SWA_acceptor_diff&#x27;, &#x27;MES-SWA_donor_alt&#x27;,\n",
       "                                  &#x27;MES-SWA_donor_diff&#x27;, &#x27;MaxEntScan_alt&#x27;,\n",
       "                                  &#x27;MaxEntScan_diff&#x27;, &#x27;GERP++_RS&#x27;,\n",
       "                                  &#x27;phastCons100way_vertebrate&#x27;,\n",
       "                                  &#x27;phyloP100way_vertebrate&#x27;]),\n",
       "                                (&#x27;pipeline-2&#x27;,\n",
       "                                 Pipeline(...\n",
       "                                                                              &#x27;tolerated&#x27;,\n",
       "                                                                              &#x27;deleterious_low_confidence&#x27;,\n",
       "                                                                              &#x27;deleterious&#x27;],\n",
       "                                                                             [&#x27;unknown&#x27;,\n",
       "                                                                              &#x27;benign&#x27;,\n",
       "                                                                              &#x27;possibly_damaging&#x27;,\n",
       "                                                                              &#x27;probably_damaging&#x27;]],\n",
       "                                                                 handle_unknown=&#x27;use_encoded_value&#x27;,\n",
       "                                                                 unknown_value=-1))]),\n",
       "                                 [&#x27;Consequence&#x27;, &#x27;IMPACT&#x27;, &#x27;SIFT&#x27;, &#x27;PolyPhen&#x27;]),\n",
       "                                (&#x27;pipeline-5&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;simpleimputer&#x27;,\n",
       "                                                  SimpleImputer(strategy=&#x27;constant&#x27;)),\n",
       "                                                 (&#x27;binaryencoder&#x27;,\n",
       "                                                  BinaryEncoder())]),\n",
       "                                 [&#x27;ORIGIN&#x27;, &#x27;BIOTYPE&#x27;])])</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"ea5bc5c4-c41b-4766-b099-9db09c2a8518\" type=\"checkbox\" ><label for=\"ea5bc5c4-c41b-4766-b099-9db09c2a8518\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">pipeline-1</label><div class=\"sk-toggleable__content\"><pre>[&#x27;MES-SWA_acceptor_alt&#x27;, &#x27;MES-SWA_acceptor_diff&#x27;, &#x27;MES-SWA_donor_alt&#x27;, &#x27;MES-SWA_donor_diff&#x27;, &#x27;MaxEntScan_alt&#x27;, &#x27;MaxEntScan_diff&#x27;, &#x27;GERP++_RS&#x27;, &#x27;phastCons100way_vertebrate&#x27;, &#x27;phyloP100way_vertebrate&#x27;]</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"38558d20-0745-4be8-9b0e-56563923138d\" type=\"checkbox\" ><label for=\"38558d20-0745-4be8-9b0e-56563923138d\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">SimpleImputer</label><div class=\"sk-toggleable__content\"><pre>SimpleImputer()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"d61c4f33-db5e-4cd5-915c-e499265a1202\" type=\"checkbox\" ><label for=\"d61c4f33-db5e-4cd5-915c-e499265a1202\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScaler</label><div class=\"sk-toggleable__content\"><pre>StandardScaler()</pre></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"2b48e78e-b475-475e-bb46-2b30f3ae2cb4\" type=\"checkbox\" ><label for=\"2b48e78e-b475-475e-bb46-2b30f3ae2cb4\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">pipeline-2</label><div class=\"sk-toggleable__content\"><pre>[&#x27;CADD&#x27;, &#x27;AF_TGP&#x27;, &#x27;Frq&#x27;, &#x27;GNOMADAF_popmax&#x27;, &#x27;Hom&#x27;, &#x27;SWEGENAF&#x27;, &#x27;MES-SWA_acceptor_alt&#x27;, &#x27;LoFtool&#x27;, &#x27;pLI_gene_value&#x27;, &#x27;SpliceAI_pred_DS_AL&#x27;, &#x27;SpliceAI_pred_DS_DG&#x27;, &#x27;SpliceAI_pred_DS_DL&#x27;]</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"2b87e5e0-6d8b-462f-84ee-119b1e2ef6b7\" type=\"checkbox\" ><label for=\"2b87e5e0-6d8b-462f-84ee-119b1e2ef6b7\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">SimpleImputer</label><div class=\"sk-toggleable__content\"><pre>SimpleImputer(fill_value=0, strategy=&#x27;constant&#x27;)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"39ea466e-0679-4e38-b413-1e17ee475690\" type=\"checkbox\" ><label for=\"39ea466e-0679-4e38-b413-1e17ee475690\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScaler</label><div class=\"sk-toggleable__content\"><pre>StandardScaler()</pre></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"50e5ddb9-a2d9-427b-8cf9-e6edb38b0cc0\" type=\"checkbox\" ><label for=\"50e5ddb9-a2d9-427b-8cf9-e6edb38b0cc0\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">missingindicator</label><div class=\"sk-toggleable__content\"><pre>[&#x27;SPIDEX&#x27;, &#x27;REVEL_score&#x27;, &#x27;SpliceAI_pred_DS_AG&#x27;, &#x27;SIFT&#x27;, &#x27;PolyPhen&#x27;]</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"1cbd221f-ae1b-48d7-bc29-9751b987cfbb\" type=\"checkbox\" ><label for=\"1cbd221f-ae1b-48d7-bc29-9751b987cfbb\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">MissingIndicator</label><div class=\"sk-toggleable__content\"><pre>MissingIndicator()</pre></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"a8dc2b38-f41b-4d75-950c-9cb131f69603\" type=\"checkbox\" ><label for=\"a8dc2b38-f41b-4d75-950c-9cb131f69603\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">pipeline-3</label><div class=\"sk-toggleable__content\"><pre>[&#x27;SPIDEX&#x27;, &#x27;REVEL_score&#x27;, &#x27;SpliceAI_pred_DS_AG&#x27;]</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"44dbf14b-947d-484f-88f7-69e621254d53\" type=\"checkbox\" ><label for=\"44dbf14b-947d-484f-88f7-69e621254d53\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">SimpleImputer</label><div class=\"sk-toggleable__content\"><pre>SimpleImputer(fill_value=0, strategy=&#x27;constant&#x27;)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"6304aba5-566a-4438-a6f5-4682e0cfd50a\" type=\"checkbox\" ><label for=\"6304aba5-566a-4438-a6f5-4682e0cfd50a\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScaler</label><div class=\"sk-toggleable__content\"><pre>StandardScaler()</pre></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"163f38d5-15a5-4af1-b241-c551ab034a4d\" type=\"checkbox\" ><label for=\"163f38d5-15a5-4af1-b241-c551ab034a4d\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">pipeline-4</label><div class=\"sk-toggleable__content\"><pre>[&#x27;Consequence&#x27;, &#x27;IMPACT&#x27;, &#x27;SIFT&#x27;, &#x27;PolyPhen&#x27;]</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"9d3a25c9-a866-4838-b797-d2304efc2809\" type=\"checkbox\" ><label for=\"9d3a25c9-a866-4838-b797-d2304efc2809\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">SimpleImputer</label><div class=\"sk-toggleable__content\"><pre>SimpleImputer(strategy=&#x27;constant&#x27;)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"7542a8ce-cdcc-42ff-bb9b-5e0bb71a5bda\" type=\"checkbox\" ><label for=\"7542a8ce-cdcc-42ff-bb9b-5e0bb71a5bda\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">OrdinalEncoder</label><div class=\"sk-toggleable__content\"><pre>OrdinalEncoder(categories=[[&#x27;intergenic_variant&#x27;, &#x27;feature_truncation&#x27;,\n",
       "                            &#x27;regulatory_region_variant&#x27;, &#x27;feature_elongation&#x27;,\n",
       "                            &#x27;regulatory_region_amplification&#x27;,\n",
       "                            &#x27;regulatory_region_ablation&#x27;,\n",
       "                            &#x27;TF_binding_site_variant&#x27;, &#x27;TFBS_amplification&#x27;,\n",
       "                            &#x27;TFBS_ablation&#x27;, &#x27;downstream_gene_variant&#x27;,\n",
       "                            &#x27;upstream_gene_variant&#x27;,\n",
       "                            &#x27;non_coding_transcript_variant&#x27;,\n",
       "                            &#x27;NMD_transcript_variant...\n",
       "                            &#x27;splice_donor_5th_base_variant&#x27;,\n",
       "                            &#x27;splice_region_variant&#x27;, &#x27;protein_altering_variant&#x27;,\n",
       "                            &#x27;missense_variant&#x27;, &#x27;inframe_deletion&#x27;, ...],\n",
       "                           [&#x27;MODIFIER&#x27;, &#x27;LOW&#x27;, &#x27;MODERATE&#x27;, &#x27;HIGH&#x27;],\n",
       "                           [&#x27;tolerated_low_confidence&#x27;, &#x27;tolerated&#x27;,\n",
       "                            &#x27;deleterious_low_confidence&#x27;, &#x27;deleterious&#x27;],\n",
       "                           [&#x27;unknown&#x27;, &#x27;benign&#x27;, &#x27;possibly_damaging&#x27;,\n",
       "                            &#x27;probably_damaging&#x27;]],\n",
       "               handle_unknown=&#x27;use_encoded_value&#x27;, unknown_value=-1)</pre></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"9f634671-d8f2-4935-8e64-8dae23491d1b\" type=\"checkbox\" ><label for=\"9f634671-d8f2-4935-8e64-8dae23491d1b\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">pipeline-5</label><div class=\"sk-toggleable__content\"><pre>[&#x27;ORIGIN&#x27;, &#x27;BIOTYPE&#x27;]</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"44f8ee91-9233-4a7b-8d53-5222492c0724\" type=\"checkbox\" ><label for=\"44f8ee91-9233-4a7b-8d53-5222492c0724\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">SimpleImputer</label><div class=\"sk-toggleable__content\"><pre>SimpleImputer(strategy=&#x27;constant&#x27;)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"7909a406-9515-403e-85fe-f576c1e40fe7\" type=\"checkbox\" ><label for=\"7909a406-9515-403e-85fe-f576c1e40fe7\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">BinaryEncoder</label><div class=\"sk-toggleable__content\"><pre>BinaryEncoder()</pre></div></div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "ColumnTransformer(transformers=[('pipeline-1',\n",
       "                                 Pipeline(steps=[('simpleimputer',\n",
       "                                                  SimpleImputer()),\n",
       "                                                 ('standardscaler',\n",
       "                                                  StandardScaler())]),\n",
       "                                 ['MES-SWA_acceptor_alt',\n",
       "                                  'MES-SWA_acceptor_diff', 'MES-SWA_donor_alt',\n",
       "                                  'MES-SWA_donor_diff', 'MaxEntScan_alt',\n",
       "                                  'MaxEntScan_diff', 'GERP++_RS',\n",
       "                                  'phastCons100way_vertebrate',\n",
       "                                  'phyloP100way_vertebrate']),\n",
       "                                ('pipeline-2',\n",
       "                                 Pipeline(...\n",
       "                                                                              'tolerated',\n",
       "                                                                              'deleterious_low_confidence',\n",
       "                                                                              'deleterious'],\n",
       "                                                                             ['unknown',\n",
       "                                                                              'benign',\n",
       "                                                                              'possibly_damaging',\n",
       "                                                                              'probably_damaging']],\n",
       "                                                                 handle_unknown='use_encoded_value',\n",
       "                                                                 unknown_value=-1))]),\n",
       "                                 ['Consequence', 'IMPACT', 'SIFT', 'PolyPhen']),\n",
       "                                ('pipeline-5',\n",
       "                                 Pipeline(steps=[('simpleimputer',\n",
       "                                                  SimpleImputer(strategy='constant')),\n",
       "                                                 ('binaryencoder',\n",
       "                                                  BinaryEncoder())]),\n",
       "                                 ['ORIGIN', 'BIOTYPE'])])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn import set_config\n",
    "\n",
    "set_config(display=\"diagram\")\n",
    "\n",
    "preprocessor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(numeric_feature_1)+len(numeric_feature_2)+len(categori_oridinal_feature_4)+len(numeric_feature_3)+len(categori_nominal_feature_5)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# train model"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LG Training set score: 0.980959\n",
      "LG Test set score: 0.981063\n"
     ]
    }
   ],
   "source": [
    "\"\"\" 6. get model\"\"\"\n",
    "\n",
    "# model: logstic regression 26s\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "logreg = LogisticRegression(max_iter=1500)   # max iteration to finish converage: max_iter=1500 , C =1\n",
    "\n",
    "\n",
    "# Define pipeline: data preprocessing + model\n",
    "pipeline = make_pipeline(preprocessor, logreg)      # preprocessor will only be used for X train automatically\n",
    "\n",
    "\n",
    "# Train model on training data\n",
    "pipeline.fit(X_train, y_train)\n",
    "\n",
    "# Evaluate model on test data\n",
    "accuracy_train_lr = pipeline.score(X_train, y_train)\n",
    "accuracy_test_lr = pipeline.score(X_test, y_test)\n",
    "\n",
    "print(\"LR Training set score: {:.6f}\".format(accuracy_train_lr)) \n",
    "print(\"LR Test set score: {:.6f}\".format(accuracy_test_lr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_LG_model.joblib']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Save the preprocessor and model: for deal with imbalance\n",
    "# dump(preprocessor, '/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_preprocessor.joblib')\n",
    "# dump(logreg, '/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_LR_model.joblib')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_LR_model_imbalance.joblib']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save the preprocessor and model: for keep imbalance\n",
    "dump(preprocessor, '/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_preprocessor_imbalance.joblib')\n",
    "dump(logreg, '/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_LR_model_imbalance.joblib')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GB Training set score: 0.989469\n",
      "GB Test set score: 0.988763\n"
     ]
    }
   ],
   "source": [
    "## try another model: Gradient boosted : 5min\n",
    "\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "gbrt = GradientBoostingClassifier(n_estimators=500,random_state=0)\n",
    "\n",
    "# Define pipeline_2: data preprocessing + model\n",
    "pipeline_2 = make_pipeline(preprocessor, gbrt)\n",
    "\n",
    "# Train model on training data\n",
    "pipeline_2.fit(X_train, y_train)\n",
    "\n",
    "# Evaluate model on test data\n",
    "accuracy_train_gb = pipeline_2.score(X_train, y_train)\n",
    "accuracy_test_gb = pipeline_2.score(X_test, y_test)\n",
    "\n",
    "print(\"GB Training set score: {:.6f}\".format(accuracy_train_gb)) \n",
    "print(\"GB Test set score: {:.6f}\".format(accuracy_test_gb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_GB_model_imbalance.joblib']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save model: the same preprocessor already saved\n",
    "\n",
    "dump(gbrt, '/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models/01_GB_model_imbalance.joblib')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SVM RBF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## try SVM RBF:  very slow\n",
    "\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "\n",
    "svc = SVC(probability=True)                  # By default, C=1 and gamma=1/n_features ; prprobability=True to make it has prob\n",
    "\n",
    "# Define pipeline_3: data preprocessing + model\n",
    "pipeline_3 = make_pipeline(preprocessor, svc)\n",
    "\n",
    "# Train model on training dataf\n",
    "pipeline_3.fit(X_train, y_train)\n",
    "\n",
    "# Evaluate model on test data\n",
    "accuracy_train_sv = pipeline_3.score(X_train, y_train)\n",
    "accuracy_test_sv = pipeline_3.score(X_test, y_test)\n",
    "\n",
    "print(\"SVM Training set score: {:.6f}\".format(accuracy_train_sv)) \n",
    "print(\"SVM Test set score: {:.6f}\".format(accuracy_test_sv))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dump(svc, '/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models//01_SVC_model_imbalance.joblib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save model: the same preprocessor already saved\n",
    "\n",
    "dump(svc, '/Users/nancy/Desktop/RS_projects/result/03_ML/01_train_model/01_featureV1/models//01_SVC_model.joblib')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8 (main, Nov 24 2022, 08:08:27) [Clang 14.0.6 ]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "c02203d1d64a20e095821dadd6770e07e454ea747f2e6a086ab8a4b96e3b68a3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
